src/java/org/apache/cassandra/utils/SystemInfo.java

```diff
@@ -0,0 +1,223 @@
+     * @see #DEFAULT_MAX_PROCESSES
+     */
+    public long getMaxProcess()
+    {
+        // this check only works on Linux systems.  Errors fall through to return default.
+        if (oshi.SystemInfo.getCurrentPlatform() == PlatformEnum.LINUX)
+        {
+            String path = format("/proc/%s/limits", getPid());
+            try
+            {
+                List<String> lines = FileUtils.readLines(new File(path));
+                for (String line : lines)
+                {
+                    if (line.startsWith("Max processes"))
+                    {
+                        String[] parts = SPACES_PATTERN.split(line);
+
+                        if (parts.length < 3)
+                            continue;
+
+                        String limit = parts[2];
+                        return "unlimited".equals(limit) ? INFINITY : Long.parseLong(limit);
+                    }
+                }
+                logger.error( "'Max processes' not found in "+path);
+            }
+            catch (Throwable t)
+            {
+                logger.error(format("Unable to read %s", path), t);
+            }
+        }
+
+        /* return the default value for non-Linux systems or parsing error.
+         * Can not return 0 as we know there is at least 1 process (this one) and
+         * -1 historically represents infinity.
+         */
+        return DEFAULT_MAX_PROCESSES;
+    }
+
+    /**
+     * @return The maximum number of open files allowd to the current process/user.
+     */
+    public long getMaxOpenFiles()
+    {
+        // ulimit -H -n
+        return si.getOperatingSystem().getCurrentProcess().getHardOpenFileLimit();
+    }
+
+    /**
+     * Gets the Virtual Memory Size (VSZ). Includes all memory that the process can access,
+     * including memory that is swapped out and memory that is from shared libraries.
+     *
+     * @return The amount of virtual memory allowed to be allocatedby the current process/user.
+     */
+    public long getVirtualMemoryMax()
+    {
+        return si.getOperatingSystem().getCurrentProcess().getVirtualSize();
+    }
+
+    /**
+     * @return The amount of swap space allocated on the system.
+     */
+    public long getSwapSize()
+    {
+        return si.getHardware().getMemory().getVirtualMemory().getSwapTotal();
+    }
+
+    /**
+     * @return the PID of the current system.
+     */
+    public long getPid()
+    {
+        return si.getOperatingSystem().getProcessId();
+    }
+
+    /**
+     * @return the Semver for the kernel version of the OS.
+     */
+    public Semver getKernelVersion()
+    { 
+        return new Semver(si.getOperatingSystem().getVersionInfo().getBuildNumber(), Semver.SemverType.LOOSE);
+    }
+
+    /**
+     * Tests if the system is running in degraded mode.
+     *
+     * @return non-empty optional with degradation messages if the system is in degraded mode, empty optional otherwise.
+     */
+    public Optional<String> isDegraded()
```

Comment:
@Claudenw I was thinking about adding some tests for this. As of now, we just don't test the logic in this method. SystemInfo might be mocked by Mockito on its methods to return whatever we want instead of oshi returning it and then we could make assertions on various corner cases and values so we might watch what degradation string we resolve here.